### ISSUE TITLE
Product List Items

### ISSUE THREAD

**Issue by josh-carter** [QubitProducts/UniversalVariable-Magento-Extension#50]
Latest version of this module makes the product list show all items no matter how many you have it set to show.

**Comment by firewizard** [comment]
are you referring to `universal_variable.listing`? If so, what do you men by "how many you have it set to show"? i just tested it and from a list of 235 products, the pages shows 25 and the uv variable holds 25 items too - the `results_count` parameter is 235, indeed, but there are only 25 items in the list.

**Comment by benjamni** [comment]
Hi @josh-carter do you have an example of this? I have had to disable the listing object on a number of sites with a similar issue

**Comment by josh-carter** [comment]
Hi @benjamni thanks for getting in touch. 

If you look at staging.publicdesire.co.uk and go to a product list, every product in that category gets listed. Also none of the OnTap Visual merchandising comes in to play.

**Comment by benjamni** [comment]
Hi @josh-carter 

We've been doing some research on this lately and it's down to third party merchandising extensions overriding the product_list block. We are trying to find a consistent method of fixing this across multiple third party extensions, but everyone seems to be doing it differently. A colleague of mine seems to remember you having a fix for your issue? Could you shed any light on how you went about this?

**Comment by rogy** [comment]
@firewizard we've been having this issue a lot lately from multiple people, I understand that the admin html configuration option for product_list is basically for this exactly problem, but a lot of these extensions don't seem to actually declare a new block name they just do:

```
class *_*_Block_Catalog_Product_List extends Mage_Catalog_Block_Product_list
```

**Comment by firewizard** [comment]
Hey @benjamni & @josh-carter - the problem resides in how the magento product_list block works. unfortunatelly all code logic is handled by the toolbar and by the product list block in the `beforeToHtml` method. see `Mage_Catalog_Block_Product_List::_beforeToHtml` and `Mage_Catalog_Block_Product_List_Toolbar::setCollection` and `Mage_Catalog_Block_Product_List_Toolbar::getPagerHtml` for more details.

this means there's only a small "window" where the correct collection can be used. and that's after the list block has generated its html. if you try to work with the product collection anywhere before the list block renders, you'll end up with inconsistent data (not to mention multiple db load calls).

take a look at how this is done in this extension. you'll notice that the `listing` variable is set in the Observer: https://github.com/QubitProducts/UniversalVariable-Magento-Extension/blob/master/app/code/community/QuBit/UniversalVariable/Model/Observer.php - which listens for `controller_action_layout_render_before`

you could use the same approach for 3rd party extensions. but it will get a bit more complicated, depending on what the extension is doing and how it's been coded. 

another workaround is to listen for the final event and manipulate the html directly, as a string. not very neat, but it will work.

**Comment by rogy** [comment]
Hi @firewizard 

This is something me and @benjamni were talking about, for the purposes of this extension, we ideally want the list of products that end up being rendered on page. Not really caring where they come from - as this is a more adaptable solution across many sites.

Do you know how we would go about grabbing the 'final' product list for a current page and I'll look into refactoring that change in?

**Comment by firewizard** [comment]
@rogy I don't understand what you mean. the product list the the extension is using is the correct list, it's the same as the one that shows up on the page (after sorting, pagination and filtering). 

if this doesn't work in some scenarios because of 3rd party extensions, the 3rd party extension is overriding something and the UV code needs to be adapted (or, even better, the 3pd party needs to be adapted/fixed). 

the code for the current listing variable is handled in the observer, as mentioned above. you can change that and load the collection in the head (together with the rest of the variables), but it's incorrect.

**Comment by rogy** [comment]
Understood @firewizard the problem is that we've had this issue now with 3 different third party extensions, and while it clearly isn't our extensions fault, we are trying to look at ways to make it work for more, as these third party merchandising extensions are rather popular.

The 3 scenarios that have occurred so far:-
 - When accessing the `$collection` in `observer.php` the third party extension stops working on the PLP
 - When accessing the `$collection` in `observer.php` all products are loaded into the PLP without merchandising from the third party extension
 - When accessing the `$collection` in `observer.php` all products are loaded into the PLP with the merchandising from third party extension

Scenario 2 has been reported 3 times, 1 and 3 one time each. Each scenario relates to a specific third party merchandising extension.

Are you saying that the only way forward, is for the UV extension to do it 'the wrong way' to try get round failures by other third parties?

**Comment by firewizard** [comment]
it's difficult to say what to do without actually knowing what the setup is. what extensions are we talking about? and what is the problem exactly, cause it's still not clear to me. I checked http://staging.publicdesire.co.uk/heels.html and the `listing` object holds correct info - 424 total count, 424 items, in the exact same order as the ones on the page...

**Comment by josh-carter** [comment]
The Problem is that it displays all products in the lsit, its set to only shoq 20 per page, its shows all 424 items, with this extension disabled it shows 20 as defined. 


Kind Regards,Josh Carter,

Magento Developer,

Webtise Ltd

**Comment by firewizard** [comment]
Josh, I'm looking at the code and this cannot happen because of this extension.
``` php
/** @var Mage_Catalog_Model_Resource_Product_Collection $collection */
$collection = $block->getLoadedProductCollection();
$listing['result_count'] = $collection->getSize();
$listing['items'] = array();
foreach($collection as $product) {
    $listing['items'][] = $uv->getProductData($product);
}
```
have you tried disabling all plugins on your store and only keeping this one to see if the problem persist? 

can you tell me what 3rd party extension you're using? of even better, can you give me access to your staging site? or I can give you a few tools to run and send me the output, for better understanding the setup.

i'm afraid that's all i can do. I'm using this extension myself and I have no issues - i can't replicate it. I'll try to install a new magento with UV only and see if I can replicate the issue

**Comment by josh-carter** [comment]
We use onTap visual merchandiser, the module Magento Enterprise comes with out the box. Fair enough if this is the problem however i would expect Magento’s first choice merchandising extension to not be the cause and rather that this extension conflicts in some way, im out at the moment but as soon as i get back to my mac i will check if disabling this solves the issue. 


Kind Regards,Josh Carter,

Magento Developer,

Webtise Ltd

**Comment by firewizard** [comment]
I had a hunch it's an EE. There might be a few (major) differences between EE and CE, and this might be one of them. It might not even be an extension but the core logic might be improved. Difficult to tell without debugging. Try adding a `return` in the observer method, to disable. If that fixes the problem, then the EE code is different and our approach doesn't work. I don't have access to an EE so I can't test that.

**Comment by josh-carter** [comment]
Its not an EE, im saying the module we use to merchandise is the one Magento uses on EE.


Kind Regards,Josh Carter,

Magento Developer,

Webtise Ltd

**Comment by rogy** [comment]
OnTap was one of the ones I had a problem with a while back, although when I was working with it, the onTap merchandiser looked like it was 'disabled' when the UV module was enabled.

The problem comes from conflicts between these third party extensions that override the product listings and the UV extension, its figuring out why is our issue.

The other module that causes a problem (scenario 2 in my list also) is the Locayta merchandising extension.

**Comment by firewizard** [comment]
I have zero experience with both of them, so without looking at code and debugging it's difficult to say where the problem is coming from. 

on top of that, the usual approach to extensions is to offer support for the magento-extension integration, not extension-extension compatibility. that usually comes down to the developer(s) installing the extensions. I'm willing to help you out on this as much as I can, but I'm just saying we might fix this and tomorrow we end up with another scenario, and so on.

**Comment by rogy** [comment]
Agreed that is usually our stance as well, its just that this has come up so many times recently.

I will follow up with you seperately and we can report back here

**Comment by rogy** [comment]
@firewizard can you send me an email to the one in my profile and I will link you up to what we're working with

**Comment by tejashp77** [comment]
Sorting and paging is not working with this module. 

i have tried in fresh magento also. anyone has solution?

**Comment by firewizard** [comment]
what magento version, what UV version? do you have a demo?

**Comment by tejashp77** [comment]
its in local..  Magento ver. 1.13.1.0 (EE)

**Comment by firewizard** [comment]
although Magento EE and CE are not that different, this has only been tested on CE 1.8+, so I can't say why it doesn't work on EE. Since I don't have any EE's that I can test with, you'll have to debug it on your own - unless someone else has any ideas or experience with EE.

**Comment by tejashp77** [comment]
but its also not working in Magento CE. Same issue is there

**Comment by firewizard** [comment]
i am using the lastest UV version with magento 1.9 and it's working fine.

**Comment by tejashp77** [comment]
UV version means extension version? its 1.0.23

**Comment by firewizard** [comment]
strange, that's what I have and things work ok. it's been coded and tested on magento 1.8 I think, so I don't see how it doesn't work with a clean install.

**Comment by tejashp77** [comment]
meanwhile this is working for me.
http://magento.codexamples.com/download/e814/r1367/qubit-universal-variable.html

**Comment by firewizard** [comment]
great. I have no idea what the differences are, but feel free to compare them

**Comment by tejashp77** [comment]
they used ur name? u don't know? i want your suggetion

**Comment by firewizard** [comment]
They credited me for my contribution to this repository. I have no idea about that extension so I cannot recommend it.

I checked the differences and I wouldn't use that plugin. First off, it's a mess, it changed a lot of things for nothing, it added a controller that it doesn't use. It added a block that was unnecessary, it changed other stuff that was coded magento-style, plus, the reason why this works for you: it took out the part about listings. I will install a new CE and test the plugin, hope to have an answer by the end of the week.

**Comment by rogy** [comment]
The link posted by @tejashp77 isn't condoned by us (Qubit) in any way.

**Comment by firewizard** [comment]
@tejashp77 you're right, the listing page is broken on mage CE 1.9.2.4 - I'll do my best to fix it this week.

**Comment by firewizard** [comment]
this is fixed in https://github.com/QubitProducts/UniversalVariable-Magento-Extension/pull/54 - waiting for @rogy to review and merge.

although not tested, the workaround solution should work fine on all version of magento.

**Comment by rogy** [comment]
@tejashp77 could you test the above? I will be doing some testing this week.
